File: C#/TripServiceLegacyKata/Trip/TripService.cs
﻿using System.Collections.Generic;
using TripServiceLegacyKata.Exception;
using TripServiceLegacyKata.User;

namespace TripServiceLegacyKata.Trip
{
    public class TripService
    {
        public List<Trip> GetTripsByUser(User.User user)
        {
            List<Trip> tripList = new List<Trip>();
            User.User loggedUser = UserSession.GetInstance().GetLoggedUser();
            bool isFriend = false;
            if (loggedUser != null)
            {
                foreach (User.User friend in user.GetFriends())
                {
                    if (friend.Equals(loggedUser))
                    {
                        isFriend = true;
                        break;
                    }
                }
                if (isFriend)
                {
                    tripList = TripDAO.FindTripsByUser(user);
                }
                return tripList;
            }
            else
            {
                throw new UserNotLoggedInException();
            }
        }
    }
}

File: C++/TripServiceLegacyKata/src/TripService.cpp
#include "TripService.h"

std::list<Trip> TripService::getTripsByUser(User user)
{
  std::list<Trip> triplist;
  User* loggedUser = getLoggedUserFromSession();
  bool isFriend = false;
  if ( loggedUser )
  {
    std::list<User>::iterator i;
    for( i = user.GetFriends().begin(); i != user.GetFriends().end(); ++i )
    {
      if ( *i == *loggedUser )
      {
        isFriend = true;
        break;
      }
    }
    if (isFriend)
    {
      triplist = findTripsByUser(user);
    }
    return triplist;
  }
  else
  {
    throw "UserNotLoggedInException";
  }
}

User* TripService::getLoggedUserFromSession() const {
  return UserSession::GetInstance()->GetLoggedUser();
} 

std::list<Trip> TripService::findTripsByUser(User user) const{
  return TripDAO::FindTripsByUser(user);
}
